cpg_workflows/query_modules/seqr_loader.py

```diff
@@ -169,58 +171,71 @@ def annotate_cohort(
     """
     Subset the MatrixTable to the provided list of samples and to variants present
     in those samples
-    @param mt_path: cohort-level matrix table from VCF.
-    @param sample_ids: samples to take from the matrix table.
-    @param out_mt_path: path to write the result.
-    @param exclusion_file: path to a file containing samples to remove from the
-            subset prior to attempting removal
+
+    Args:
+        mt_path (str): cohort-level matrix table from VCF.
+        sample_ids (list[str]): samples to take from the matrix table.
+        out_mt_path (str): path to write the result.
+        exclusion_file (str, optional): path to a file containing samples to remove from the
+                                        subset prior to extracting
     """
 
-    mt = hl.read_matrix_table(str(mt_path))
+    mt = hl.read_matrix_table(mt_path)
 
-    sample_ids = set(sample_ids)
+    unique_sids: set[str] = set(sample_ids)
 
     # if an exclusion file was passed, remove the samples from the subset
     # this executes in a query command, by execution time the file should exist
     if exclusion_file:
         with hl.hadoop_open(exclusion_file) as f:
             exclusion_ids = set(f.read().splitlines())
         logging.info(f'Excluding {len(exclusion_ids)} samples from the subset')
-        sample_ids -= exclusion_ids
+        unique_sids -= exclusion_ids
 
     mt_sample_ids = set(mt.s.collect())
 
-    sample_ids_not_in_mt = sample_ids - mt_sample_ids
-    if sample_ids_not_in_mt:
+    if sample_ids_not_in_mt := unique_sids - mt_sample_ids:
         raise Exception(
-            f'Found {len(sample_ids_not_in_mt)}/{len(sample_ids)} samples '
-            f'in the subset set that do not matching IDs in the variant callset.\n'
+            f'Found {len(sample_ids_not_in_mt)}/{len(unique_sids)} IDs in the requested subset not in the callset.\n'
             f'IDs that aren\'t in the callset: {sample_ids_not_in_mt}\n'
             f'All callset sample IDs: {mt_sample_ids}',
         )
 
-    logging.info(f'Found {len(mt_sample_ids)} samples in mt, subsetting to {len(sample_ids)} samples.')
+    # work out the proportion of the original sample count that we'll keep
+    downsample_ratio = len(mt_sample_ids) // len(unique_sids)
+    # calculate the proportional number of partitions to generate
+    downsample_partitions = int(mt.n_partitions() * downsample_ratio)
+
+    logging.info(f'Found {len(mt_sample_ids)} samples in mt, subsetting to {len(unique_sids)} samples.')
+    logging.info(f'Re-reading the MT with {downsample_partitions} partitions, reduced from {mt.n_partitions()}')
+    # re-read instead of a repartition/naive_coalesce
+    # naive coalesce could produce really unbalanced partitions, and repartitioning might be more expensive
+    # "it's basically free to read with different partitioning than you wrote with"
```

Comment:
I don't believe this - in Talos both running on a high-partition dataset and reading that partitioned dataset with a different number of partitions cause runtime failures. Reading a dataset, writing it back out with a different number of partitions, then reading that in again worked OK..

I expect QOB can handle it, but local-pyspark'y Hail cannot